File: solutions/_394_decode_string/solution.java

package _394_decode_string;

import java.util.Stack;

class Solution {
    public String decodeString(String s) {
        var stack = new Stack<String>();
        var res = new StringBuilder();
        String k = "";

        for (var ch : s.toCharArray()) {
            if (Character.isDigit(ch)) {
                k += ch;
            } else if (ch == '[') {
                stack.push(res.toString());
                stack.push(k);
                res = new StringBuilder();
                k = "";
            } else if (ch == ']') {
                int times = Integer.parseInt(stack.pop());
                var prevRes = new StringBuilder(stack.pop());
                while (times-- > 0) {
                    prevRes.append(res);
                }
                res = prevRes;
            } else {
                res.append(ch);
            }
        }

        return res.toString();
    }
}

/*
 * https://leetcode.com/problems/decode-string/
 */

File: solutions/_394_decode_string/solution.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        res = ""
        k = ""

        for char in s:
            if char.isdigit():
                k += char
            elif char == "[":
                stack.append([k, res])
                k = ""
                res = ""
            elif char == "]":
                prev_k, prev_res = stack.pop()
                res = prev_res + int(prev_k) * res
            else:
                res += char

        return res


# https://leetcode.com/problems/decode-string/
